src/hotspot/share/gc/shenandoah/shenandoahFreeSet.cpp

```diff
@@ -688,67 +918,187 @@ void ShenandoahFreeSet::rebuild() {
+      young_reserve = _heap->get_young_evac_reserve();
+      old_reserve = _heap->get_promoted_reserve() + _heap->get_old_evac_reserve();
+    } else {
+      // We are rebuilding at end of GC, so we set aside budgets specified on command line (or defaults)
+      young_reserve = (_heap->young_generation()->max_capacity() * ShenandoahEvacReserve) / 100;
+      old_reserve = MAX2((_heap->old_generation()->max_capacity() * ShenandoahOldEvacReserve) / 100,
+                         ShenandoahOldCompactionReserve * ShenandoahHeapRegion::region_size_bytes());
+    }
   }
-
+  reserve_regions(young_reserve, old_reserve);
   recompute_bounds();
   assert_bounds();
+  log_status();
 }
 
-void ShenandoahFreeSet::reserve_regions(size_t to_reserve) {
+void ShenandoahFreeSet::reserve_regions(size_t to_reserve, size_t to_reserve_old) {
   size_t reserved = 0;
-
   for (size_t idx = _heap->num_regions() - 1; idx > 0; idx--) {
-    if (reserved >= to_reserve) break;
-
-    ShenandoahHeapRegion* region = _heap->get_region(idx);
-    if (_mutator_free_bitmap.at(idx) && can_allocate_from(region)) {
-      _mutator_free_bitmap.clear_bit(idx);
-      _collector_free_bitmap.set_bit(idx);
-      size_t ac = alloc_capacity(region);
-      _capacity -= ac;
-      reserved += ac;
-      log_debug(gc, free)("  Shifting Region " SIZE_FORMAT " (Free: " SIZE_FORMAT "%s, Used: " SIZE_FORMAT "%s) to collector free set",
-                          idx, byte_size_in_proper_unit(region->free()), proper_unit_for_byte_size(region->free()),
-                               byte_size_in_proper_unit(region->used()), proper_unit_for_byte_size(region->used()));
+    ShenandoahHeapRegion* r = _heap->get_region(idx);
+    if (_mutator_free_bitmap.at(idx) && (alloc_capacity(r) > 0)) {
+      assert(!r->is_old(), "mutator_is_free regions should not be affiliated OLD");
+      // OLD regions that have available memory are already in the old_collector free set
+      if ((_old_capacity < to_reserve_old) && (r->is_trash() || (r->affiliation() == ShenandoahAffiliation::FREE))) {
+        clear_mutator_free(idx);
+        set_old_collector_free(idx);
+        size_t ac = alloc_capacity(r);
+        _capacity -= ac;
+        _old_capacity += ac;
+        log_debug(gc, free)("  Shifting region " SIZE_FORMAT " from mutator_free to old_collector_free", idx);
+      } else if (reserved < to_reserve) {
+        // Note: In a previous implementation, regions were only placed into the survivor space (collector_is_free) if
+        // they were entirely empty.  I'm not sure I understand the rational for that.  That alternative behavior would
+        // tend to mix survivor objects with ephemeral objects, making it more difficult to reclaim the memory for the
+        // ephemeral objects.  It also delays aging of regions, causing promotion in place to be delayed.
+        clear_mutator_free(idx);
+        set_collector_free(idx);
```

Comment:
Is there a situation when a region can be both in collector and in mutator free sets?